### PR TITLE
Add Details in ArticleForm

### PR DESCRIPTION
**Pull request by sirriah**
Details - first step

### CODE REVIEW

File: components/ArticleForm.js [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import { useState, useMemo } from 'react';
 import Image from 'next/image';
-import { useRouter } from 'next/router';
+// import { useRouter } from 'next/router';
```
Comment (GLObus303) [inline components/ArticleForm.js]: Uncomment :F

File: components/ArticleForm.js [hunk 2]
```diff
@@ -41,6 +43,8 @@ export const ArticleForm = ({
     featuredImage || '',
   );
```
Comment (GLObus303) [inline components/ArticleForm.js]: No need for new lines between useStates

File: components/ArticleForm.js [hunk 3]
```diff
@@ -197,6 +211,12 @@ export const ArticleForm = ({
             rows="4"
             defaultValue={altForFeaturedImage || ''}
           />
+          <label className="headline--4">Detaily</label>
+          <Details
+            id="details"
+            detailsDataCallback={handleDetailsData}
```
Comment (GLObus303) [inline components/ArticleForm.js]: Consider this: 
```suggestion
            detailsDataCallback={setDetailsState}
```

File: components/Details.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import { useState, useEffect } from 'react';
+
+import { transformDetailsToFields } from './../utils/transformData';
+import { FormInput } from './Form/FormInput';
+export const Details = ({ detailsDataCallback, defaultValue }) => {
```
Comment (GLObus303) [inline components/Details.js:5]: new line before component

File: components/Details.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            <button
+              className="bg-red-700 p-2 text-white"
+              onClick={(event) => handleDeleteFields(index, event)}
+            >
+              smazat
+            </button>
+          </div>
+        </li>
+      ))}
+      <button
```
Comment (GLObus303) [inline components/Details.js:67]: type="button"

File: components/Details.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    transformDetailsToFields(defaultValue) || [],
+  );
+
+  const handleAddFields = (event) => {
+    event.preventDefault();
+    setDetailFields([...detailFields, { fieldName: '', fieldContent: '' }]);
+  };
+
+  const handleDeleteFields = (index, event) => {
+    event.preventDefault();
```
Comment (GLObus303) [inline components/Details.js]: If the button's type is `button`. We can get rid of the `preventDefault()`

File: components/Details.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
+import { useState, useEffect } from 'react';
+
+import { transformDetailsToFields } from './../utils/transformData';
+import { FormInput } from './Form/FormInput';
+export const Details = ({ detailsDataCallback, defaultValue }) => {
+  const [detailFields, setDetailFields] = useState(
+    transformDetailsToFields(defaultValue) || [],
```
Comment (GLObus303) [inline components/Details.js]: Consider either adding this array [{fieldName, fieldContent}, ... ] directly to firebase

File: components/Details.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    event.preventDefault();
+    setDetailFields([...detailFields, { fieldName: '', fieldContent: '' }]);
+  };
+
+  const handleDeleteFields = (index, event) => {
+    event.preventDefault();
+    const data = [...detailFields];
+    data.splice(index, 1);
+    setDetailFields(data);
+  };
```
Comment (GLObus303) [inline components/Details.js]:
```suggestion
  const handleDeleteFields = (indexToRemove, event) => {
    event.preventDefault();
    
    const newData = detailFields.filter((_, index) => index === indexToRemove);
    
    setDetailFields(newData);
  };
```

File: components/Details.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+  const handleFormChange = (index, event) => {
+    const data = [...detailFields];
+    data[index][event.target.name] = event.target.value;
+    setDetailFields(data);
+  };
```
Comment (GLObus303) [inline components/Details.js]: In general, if you want to clone something not to MUTATE the original structure. you have two approaches

const newArr = [ ...oldArray ]

or 

const newArray = oldArray.slice(0)

it is equivalent

Comment (GLObus303) [reply components/Details.js]:
```suggestion
  const handleFormChange = (index, { target }) => {
    const newData = newArray.map((item) => item.fieldName === target.name ? { ...item, fieldName: target.value} : item)
    
    setDetailFields(data);
  };
```

File: components/Details.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+  useEffect(() => {
```
Comment (GLObus303) [inline components/Details.js]: Component should follow order of logical groups as follows

1. selectors/data/router
2. useState
3. useEffect
4. handler
5. dataTransformation for JSX

File: components/Details.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+            name="fieldContent"
+            labelClassName="mt-6 mb-1 block text-sm text-stone-700"
+            className="mb-4 block w-full border-b-2 border-stone-300 bg-white p-2"
+            value={fields.fieldContent}
+            onChange={(event) => handleFormChange(index, event)}
+            id={`fieldContent[${index}]`}
+          />
+
+          <div className="flex">
+            <button
```
Comment (GLObus303) [inline components/Details.js:57]: type="button"

File: pages/article/[articleLink].js [hunk 12]
```diff
@@ -111,12 +111,12 @@ const Article = () => {
                   <h3 className="py-3 font-yrsa text-lg font-medium uppercase tracking-widest shadow-neutral-900">
                     Detaily
                   </h3>
-                  {Object.entries(details).map(({ detailId, value }) => (
-                    <div key={detailId} className="bg-stone-100 p-2">
+                  {Object.entries(details).map((item) => (
```
Comment (GLObus303) [inline pages/article/[articleLink].js]: As I have pointed in multiple comments in multiple PR :D. Consider doing 
```suggestion
                  {Object.entries(details).map(([detailId, value]) => (
```

or any other suiting names

File: utils/transformData.js [hunk 13]
```diff
@@ -58,6 +58,8 @@ export const parseKeywords = (keywordsString) => {
 
     return trimmedArray.filter((item) => item.length > 0);
   }
+
+  return '';
```
Comment (GLObus303) [inline utils/transformData.js]: Do not change returned structure/type. Return either null and handle it, or return empty array

File: utils/transformData.js [hunk 14]
```diff
@@ -78,3 +80,36 @@ export const transformFirstCharToUpperCase = (string) => {
...
+export const transformDetailsToFields = (formDataDetailsArray) => {
+  if (formDataDetailsArray) {
+    const array = Object.entries(formDataDetailsArray).reduce((acc, item) => {
+      const helpArray = {};
+      helpArray.fieldName = item[0];
+      helpArray.fieldContent = item[1];
+      acc.push(helpArray);
+
+      return acc;
+    }, []);
```
Comment (GLObus303) [inline utils/transformData.js]: const transformDetailsToFields = (formDataDetailsArray) => {
  if (!formDataDetailsArray) {
    return []
  }

   return Object.entries(formDataDetailsArray).map(([fieldName, fieldContent]) => ({fieldName, fieldContent}));
};

File: utils/transformData.js [hunk 15]
```diff
@@ -78,3 +80,36 @@ export const transformFirstCharToUpperCase = (string) => {
...
+
+    return array;
+  }
+
+  return '';
+};
```
Comment (GLObus303) [inline utils/transformData.js]: const transformDetailsToFields = (formDataDetailsArray) => {
  if (!formDataDetailsArray) {
    return []
  }

   return Object.entries(formDataDetailsArray).map(([fieldName, fieldContent]) => ({fieldName, fieldContent}));
};

File: utils/transformData.js [hunk 16]
```diff
@@ -78,3 +80,36 @@ export const transformFirstCharToUpperCase = (string) => {
...
+      acc[fieldName] = '';
+    }
+
+    acc[fieldName] = item.fieldContent;
+
+    return acc;
+  }, {});
+
+  return array;
+};
```
Comment (GLObus303) [inline utils/transformData.js]:
```suggestion
export const transformDetailsToObject = (formDataDetailsArray) => formDataDetailsObject.reduce((acc, { fieldName, fieldContent }) => {
    acc[fieldName] = { fieldName: fieldName || '', fieldContent};

    return acc;
  }, {});
```

**Review by GLObus303 — APPROVED** [review]

File: components/Details.js (hunk 8 above)
Comment (sirriah) [reply components/Details.js]: Line 18: I think, that there should be a !==

File: components/Details.js (hunk 9 above)
Comment (sirriah) [reply components/Details.js]: I reused your code

```

  const handleFormChange = (index, { target }) => {
    const newData = detailFields.map((item, i) =>
      index === i ? { ...item, [target.name]: target.value } : item,
    );

    setDetailFields(newData);
  };
```

**Pull request merged**